lib/matplotlib/layout_engine.py

```diff
@@ -0,0 +1,230 @@
+to deal with colorbars, so having a dedicated property of the figure to
+querry is useful.
+
+Matplotlib has two native layout engines, ``tight_layout`` and
+``constrained_layout``, which are implimented using this formalism
+as `.tight_layout_engine` and `.constrained_layout_engine`.  While layout
+engines tend to be complicated, users and downstream libraries can now create
+their own layout engine and connect it to a figure.
+"""
+
+import matplotlib as mpl
+import matplotlib._api as _api
+
+from matplotlib._constrained_layout import do_constrained_layout
+from matplotlib.tight_layout import (get_subplotspec_list,
+                                     get_tight_layout_figure)
+from matplotlib.backend_bases import _get_renderer
+from contextlib import nullcontext
+
+
+class LayoutEngine():
+    """
+    Base class for matplotlib layout engines.
+
+    A layout engine can be passed to a figure at instantiation or
+    at any time with `~.figure.Figure.set_layout_engine`.  However, note
+    note that layout engines affect the creation of colorbars, so
+    `~.figure.Figure.set_layout_engine` should be called before any
+    colorbars are created.
+
+    Once attached to a figure, the layout engine ``execute`` function
+    is called at draw time by `~.figure.Figure.draw`, providing a special
+    draw-time hook.
+
+    Currently, there are two properties of ``LayoutEngine`` classes that
+    are consulted while manipulating the figure.  ``engine.colorbar_gridspec``
+    tells `.Figure.colorbar` whether to make the axes using the gridspec
+    method (see `.colorbar.make_axes_gridspec`) or not
+    (see `.colorbar.make_axes`); for the native layout engines
+    `.constrained_layout_engine` sets this to *False*.  The second property
+    is ``engine.adjust_compatible`` that stops `.Figure.subplots_adjust` from
+    being run if it is not compatible with the layout engine
+    (`.constrained_layout_engine` sets this to *False* also).
+    """
+    def __init__(self, **kwargs):
+        self._figure = None
+        self.colorbar_gridspec = True
+        self._default = {}
+        self._keys = []
+        self._params = {}
+        self.adjust_compatible = False
+
+    def set_params(self, **kwargs):
+        raise(NotImplementedError)
+
+    def get_params(self):
+        return self._params
+
+    def execute(self):
+        raise(NotImplementedError)
+
+    def set_figure(self, figure):
+        """
+        Set the figure for the layout manager.
+        """
+        if self._figure == figure:
+            return
+        if self._figure is None:
+            self._figure = figure
+        else:
+            raise ValueError(
+                'Layout managers cannot be attached to more than one figure')
+
+    def get_figure(self):
+        return self._figure
+
+
+class tight_layout_engine(LayoutEngine):
+    """
+    Impliments the ``tight_layout`` geometry management.
+    """
+
+    def __init__(self, **kwargs):
```

Comment:
The default pad should be defined/documented here, rather than being all the way down in the implementation of set_params.  Also perhaps make the signature explicit, i.e. `def __init__(self, *, pad=1.08, h_pad=None, w_pad=None, rect=(0, 0, 1, 1))`

Reply:
Agreed.  I think I did this right.